### PR TITLE
[#233 #17]: Rethink configuration storage at project level

### PR DESCRIPTION
**Pull request by MacFJA**
- Read configuration from .phive and current working directory (phars + auth)
- Update skel command
- Add warning in output if both configuration exist
- Add project level repositories
- Copy phive.xml to .phive/phars.xml

### CODE REVIEW

**Comment by MacFJA** [comment]
I still need to add more test

File: src/services/migration/MigrationService.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * MigrationService constructor.
+     */
+    public function __construct(MigrationFactory $factory) {
+        $this->factory = $factory;
+    }
+
+    public function ensureFitness(): void {
+        $this->runMigrations(true);
```
Comment (theseer) [inline src/services/migration/MigrationService.php:16]: This is imho bad design: It's unexpected that a test for fitness actually does change state.

We should separate the running of migrations from checking the state.

File: src/shared/config/AuthXmlConfigFileLocator.php [hunk 2]
```diff
@@ -2,24 +2,37 @@
...
+    public function __construct(Environment $environment, Config $config, Cli\Output $output) {
         $this->environment = $environment;
         $this->config      = $config;
+        $this->output      = $output;
     }
 
     public function getFile(bool $global): \PharIo\FileSystem\Filename {
         if ($global) {
-            return $this->config->getHomeDirectory()->file(self::FILENAME);
+            return $this->config->getHomeDirectory()->file('auth.xml');
```
Comment (theseer) [inline src/shared/config/AuthXmlConfigFileLocator.php:22]: Same again: let's move the `file()` call into the config, e.g. `$this->config->getHomeAuthFile()` ?

File: src/shared/exceptions/ExtensionsMissingException.php [hunk 3]
```diff
@@ -8,8 +8,8 @@ class ExtensionsMissingException extends \Exception {
     /**
      * ExtensionsMissingException constructor.
      */
-    public function __construct(array $missing) {
-        $this->missing = $missing;
+    public function __construct(array $failed) {
+        $this->missing = $failed;
         parent::__construct();
```
Comment (theseer) [inline src/shared/exceptions/ExtensionsMissingException.php:13]: ??

File: src/Factory.php [hunk 4]
```diff
@@ -249,7 +258,7 @@ public function getPharRegistry(): PharRegistry {
         if ($this->registry === null) {
             $this->registry = new PharRegistry(
                 new XmlFile(
-                    $this->getConfig()->getHomeDirectory()->file('/phars.xml'),
+                    $this->getConfig()->getHomeDirectory()->file('/database.xml'),
```
Comment (theseer) [inline src/Factory.php:261]: I think I like `registry.xml` better. What do you think?

File: tests/unit/shared/config/fixtures/doubleConfig/.phive/auth.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" ?>
+<!--
+    PHIVE AUTH CONFIGURATION SKELETON FILE
+
+    This file was automatically generated by PHIVE %%VERSION%% on %%DATE%%
+-->
+<auth xmlns="https://phar.io/phive-auth">
```
Comment (theseer) [inline tests/unit/shared/config/fixtures/doubleConfig/.phive/auth.xml:7]: Let's fix the namespace: `https://phar.io/phive/auth`

File: tests/unit/shared/config/fixtures/doubleConfig/phive-auth.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" ?>
+<!--
+    PHIVE AUTH CONFIGURATION SKELETON FILE
+
+    This file was automatically generated by PHIVE %%VERSION%% on %%DATE%%
+-->
+<auth xmlns="https://phar.io/phive-auth">
```
Comment (theseer) [inline tests/unit/shared/config/fixtures/doubleConfig/phive-auth.xml:7]: Same here :)

File: src/shared/exceptions/ExtensionsMissingException.php (hunk 3 above)
Comment (MacFJA) [reply src/shared/exceptions/ExtensionsMissingException.php:13]: My bad, I give too much power to my IDE! (a.k.a. "refactoring tool go wild" 😂)

File: src/Factory.php (hunk 4 above)
Comment (MacFJA) [reply src/Factory.php:261]: The only issue I have with `repository.xml` is how close it is to `repositories.xml`.
The two have a different purpose:
 - store the list of all installed phars for the first
 - list known phars that can be downloaded (and installed) for the second

To be honest, I'm not sure that `database.xml` fit either, it's a bit too abstract.
The more I think about it, the more `installed.xml` seem relevant.

Comment (theseer) [reply src/Factory.php:261]: I suggested `registry.xml` - not `repository.xml' ;-)

File: src/commands/migrate/MigrateCommandConfig.php [hunk 7]
```diff
@@ -10,7 +10,7 @@ public function __construct(Cli\Options $options) {
         $this->cliOptions = $options;
     }
 
-    public function showList(): bool {
+    public function showStatus(): bool {
         return $this->cliOptions->hasOption('list');
```
Comment (theseer) [inline src/commands/migrate/MigrateCommandConfig.php:14]: 'list' => 'status' ?

**Comment by theseer** [comment]
Very nice :)

Looks like we can almost merge this and finally make a release :+1:

File: src/services/migration/HomePhiveXmlMigration.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    public function migrate(): void {
+        if (!$this->canMigrate()) {
+            throw new MigrationException();
+        }
+
+        $old = $this->config->getHomeDirectory()->file('phive.xml');
+        $new = $this->config->getHomeDirectory()->file('global.xml');
+
+        $new->putContent($old->read()->getContent());
+        $old->delete();
```
[the remark this answers is not in the document]
Comment (MacFJA) [reply src/services/migration/HomePhiveXmlMigration.php:30]: I need the PR phar-io/filesystem#2 to be merge to change that

[the remark this answers is not in the document]
Comment (theseer) [reply src/services/migration/HomePhiveXmlMigration.php:30]: Already at it :)

**Comment by theseer** [comment]
Released `phar-io/filesystem` 1.0.0, released `phar-io/version` 3.0.0.

Now a test in `phar-io/manifest` fails, so I probably introduced a BC break in 3.0.0. Once all that works again, I'll update the dependencies for phive.

**Comment by theseer** [comment]
Dependencies updated, seems to work so far ;)

**Comment by theseer** [comment]
Thanks! I'll look at it tomorrow.

File: src/services/migration/HomeAuthXmlMigration.php [hunk 9]
```diff
@@ -36,7 +36,7 @@ public function migrate(): void {
         $newContent = \str_replace('https://phar.io/phive-auth', 'https://phar.io/auth', $oldContent);
         $new->putContent($newContent);
 
-        $this->output->writeText('Migration of global authentication configuration almost finish.');
+        $this->output->writeText('Migration of global authentication configuration almost finish. ');
```
Comment (theseer) [inline src/services/migration/HomeAuthXmlMigration.php:39]: Two problems:

- English grammer: It probably should be "finished"
- I'm not sure the sentence is useful as is. Maybe it should be merged with the question below?

File: src/services/migration/HomePharsXmlMigration.php [hunk 10]
```diff
@@ -34,7 +34,7 @@ public function migrate(): void {
 
         $new->putContent($old->read()->getContent());
 
-        $this->output->writeText('Migration of global authentication configuration almost finish.');
+        $this->output->writeText('Migration of list of installed phars almost finish. ');
```
Comment (theseer) [inline src/services/migration/HomePharsXmlMigration.php:37]: Same here.

File: src/services/migration/HomePhiveXmlMigration.php [hunk 11]
```diff
@@ -34,7 +34,7 @@ public function migrate(): void {
 
         $new->putContent($old->read()->getContent());
 
-        $this->output->writeText('Migration of global phive configuration almost finish.');
+        $this->output->writeText('Migration of global phive configuration almost finish. ');
```
Comment (theseer) [inline src/services/migration/HomePhiveXmlMigration.php:37]: :)

File: src/services/migration/ProjectAuthXmlMigration.php [hunk 12]
```diff
@@ -42,7 +42,7 @@ public function migrate(): void {
         $newContent = \str_replace('https://phar.io/phive-auth', 'https://phar.io/auth', $oldContent);
         $new->putContent($newContent);
 
-        $this->output->writeText('Migration of authentication configuration almost finish.');
+        $this->output->writeText('Migration of authentication configuration almost finish. ');
```
Comment (theseer) [inline src/services/migration/ProjectAuthXmlMigration.php:45]: and last but not least..

File: tests/unit/services/migration/HomeAuthXmlMigrationTest.php [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        @\unlink(__DIR__ . '/tmp/auth.xml');
+        @\unlink(__DIR__ . '/tmp/phive-auth.xml');
+        @\unlink(__DIR__ . '/tmp/phive-auth.xml.backup');
+        parent::tearDown();
+    }
+
+    /**
+     * phive-auth.xml exists, and auth.xml exists
+     */
+    public function testInError(): void {
```
Comment (theseer) [inline tests/unit/services/migration/HomeAuthXmlMigrationTest.php:23]: I like the test implementation but why put the actually useful test name into a docblock and use a rather generic method name?

I think we should rename the test to something like `testMigrationFailsWhenOldAndNewAuthFileExist`.

Same goes for the other test methods.

File: tests/unit/shared/version/GitHubVersionTest.php [hunk 14]
```diff
@@ -16,17 +16,17 @@ class GitHubVersionTest extends TestCase {
     public function testRemovesVersionPrefix($prefix): void {
         $version = new GitHubVersion($prefix . '2.9.0');
         $this->assertSame('2.9.0', $version->getVersionString());
-        $this->assertEquals(new VersionNumber('2'), $version->getMajor());
```
Comment (theseer) [inline tests/unit/shared/version/GitHubVersionTest.php:19]: This file as been removed. You can drop it.

File: src/services/migration/ProjectAuthXmlMigration.php (hunk 12 above)
Comment (MacFJA) [reply src/services/migration/ProjectAuthXmlMigration.php]: At least I'm constant in what I'm write 😃

**Comment by theseer** [comment]
Looks all good.

This change became so big, I probably missed something. :)

**Comment by MacFJA** [comment]
Yes, it goes from a not so big PR to a very big PR in no time. 😅 
But we made lots of additional changes (like refactor the `Config` class, maybe we should have made a different PR for that, to keep this one smaller 🤔)

I hope that I don't left some mistakes 🤞

**Pull request merged**